rcl/test/rcl/test_graph.cpp

```diff
@@ -1224,75 +1225,209 @@ TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_graph_query_functio
+ * service create/destroy, client create/destroy
+ * Other node added/removed
  *
  * Note: this test could be impacted by other communications on the same ROS Domain.
  */
-TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_graph_guard_condition_topics) {
+TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_graph_guard_condition_trigger_check) {
   rcl_ret_t ret;
-  // Create a thread to sleep for a time, then create a publisher, sleep more, then a subscriber,
-  // sleep more, destroy the subscriber, sleep more, and then destroy the publisher.
-  std::promise<bool> topic_changes_promise;
-  std::thread topic_thread(
-    [this, &topic_changes_promise]() {
-      // sleep
-      std::this_thread::sleep_for(std::chrono::milliseconds(200));
-      // create the publisher
-      rcl_publisher_t pub = rcl_get_zero_initialized_publisher();
-      rcl_publisher_options_t pub_ops = rcl_publisher_get_default_options();
-      rcl_ret_t ret = rcl_publisher_init(
-        &pub, this->node_ptr, ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes),
-        "/chatter_test_graph_guard_condition_topics", &pub_ops);
-      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-      // sleep
-      std::this_thread::sleep_for(std::chrono::milliseconds(200));
-      // create the subscription
-      rcl_subscription_t sub = rcl_get_zero_initialized_subscription();
-      rcl_subscription_options_t sub_ops = rcl_subscription_get_default_options();
-      ret = rcl_subscription_init(
-        &sub, this->node_ptr, ROSIDL_GET_MSG_TYPE_SUPPORT(test_msgs, msg, BasicTypes),
-        "/chatter_test_graph_guard_condition_topics", &sub_ops);
-      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-      // sleep
-      std::this_thread::sleep_for(std::chrono::milliseconds(200));
-      // destroy the subscription
-      ret = rcl_subscription_fini(&sub, this->node_ptr);
-      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-      // sleep
-      std::this_thread::sleep_for(std::chrono::milliseconds(200));
-      // destroy the publication
-      ret = rcl_publisher_fini(&pub, this->node_ptr);
-      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
-      // notify that the thread is done
-      topic_changes_promise.set_value(true);
-    });
-  // Wait for the graph state to change, expecting it to do so at least 4 times,
-  // once for each change in the topics thread.
+  std::chrono::nanoseconds time_to_sleep = std::chrono::milliseconds(400);
+
+  // Create new context
+  rcl_init_options_t init_options = rcl_get_zero_initialized_init_options();
+  ret = rcl_init_options_init(&init_options, rcl_get_default_allocator());
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    EXPECT_EQ(RCL_RET_OK, rcl_init_options_fini(&init_options)) << rcl_get_error_string().str;
+  });
+
+  // Test in new ROS domain
+  ret = rcl_init_options_set_domain_id(&init_options, 66);
```

Comment:
> AFAIU, some of the CI machines have an assigned domain id to avoid collisions with other machines in the network, and changing the number can cause random failures.

Get it.  
But for this test, we need new domain to do test. How to choose suitable domain ID?  
After checking other test codes, I find domain ID "42" are used in below test codes. 
Whether can I use it ?

https://github.com/ros2/rcl/blob/3a9711180a240e07578ce9a9522a9bf25d6ac445/rcl/test/rcl/test_node.cpp#L119-L121

https://github.com/ros2/rcl/blob/3a9711180a240e07578ce9a9522a9bf25d6ac445/rcl/test/rcl/test_domain_id.cpp#L26-L27

Reply:
> But for this test, we need new domain to do test. How to choose suitable domain ID?

Why a different domain id is needed?

Reply:
If use old domain, previous test will affect graph change.  So use new domain for clear test environment.  
Is there any way which can clear existing graph change ?    
If yse, we can use existing domain.

Reply:
> Is there any way which can clear existing graph change ?

You could wait repeatedly with a timeout until the graph guard condition is not more ready, that should "clear" previous graph changes.

Reply:
> You could wait repeatedly with a timeout until the graph guard condition is not more ready, that should "clear" previous graph changes.

Okay. I will try it.